### PR TITLE
Introduce a `GetItemId` enum

### PR DESCRIPTION
**Pull request by AngheloAlf**
As proposed on discord I made a `GetItemId` enum, which represents the entries from the `item_table` table
This PR includes an enums for both C and Python code

I tried to replace as many instances of GI raw values and replace them with the enum values. If I missed any please tell me!

There's no "compiled" files diff (bundle.o, asm_symbols.txt, etc) because the enum doesn't impact the codegen in any way.

There are a few places where the code had comments indicating to which GI the raw value corresponds to. This is no longer necessary with the enums, but I left them in case you guys still want them, if not I can remove them.
```c
        case 0: return GI_GORONS_BRACELET; // Goron Bracelet
        case 1: return GI_SILVER_GAUNTLETS; // Silver Gauntlets
        default: return GI_GOLD_GAUNTLETS; // Gold Gauntlets
```

## Advantages of using an enum

- Improves readability and searchability.
- No need for comments indicating to which item a magic number corresponds to
  - i.e `return GI_GORONS_BRACELET;` doesn't need a comment like `return 0x54; // Goron Bracelet`
- Using an enum for the GI items helps a lot for cases when you have a branch/PR that adds new GIs and another PR that adds GIs gets merged to upstream, values for the table would automatically not overlap between them, making merging upstream less error-prone.

## Disadvantages of using an enum
- The `item_table` ends up with way too much spacing. I guess I could try reducing the spacing by category or something if preferred.

## Future work

- Enums themselves can be used as types, improving readability even more.

For example (click on details to expand):

<details>
The following code:

```c
    // example 1
    uint16_t resolved_item_id = resolve_upgrades(override);
    switch (resolved_item_id) {
        case GI_OCARINA_FAIRY: { // Fairy Ocarina
            z64_file.items[Z64_SLOT_OCARINA] = 0x07;
            break;
        }
        case GI_OCARINA_OF_TIME: { // Ocarina of Time
            z64_file.items[Z64_SLOT_OCARINA] = 0x08;
            break;
        }
    }

    // example 2
uint16_t strength_upgrade(z64_file_t *save, override_t override) {
    switch ((override.value.base.player == PLAYER_ID || !MW_PROGRESSIVE_ITEMS_ENABLE) ? save->strength_upgrade : MW_PROGRESSIVE_ITEMS_STATE[override.value.base.player].strength) {
        case 0: return GI_GORONS_BRACELET; // Goron Bracelet
        case 1: return GI_SILVER_GAUNTLETS; // Silver Gauntlets
        default: return GI_GOLD_GAUNTLETS; // Gold Gauntlets
    }
}

    // example 3
item_row_t *get_item_row(uint16_t item_id) {
    if (item_id >= GI_RANDO_MAX) {
        return NULL;
    }
    item_row_t *item_row = &(item_table[item_id]);
    if (item_row->base_item_id == 0) {
        return NULL;
    }
    return item_row;
}
```

would becomes this:

```c
    // example 1
    GetItemID resolved_item_id = resolve_upgrades(override);
    switch (resolved_item_id) {
        case GI_OCARINA_FAIRY: { // Fairy Ocarina
            z64_file.items[Z64_SLOT_OCARINA] = 0x07;
            break;
        }
        case GI_OCARINA_OF_TIME: { // Ocarina of Time
            z64_file.items[Z64_SLOT_OCARINA] = 0x08;
            break;
        }
    }

    // example 2
GetItemID strength_upgrade(z64_file_t *save, override_t override) {
    switch ((override.value.base.player == PLAYER_ID || !MW_PROGRESSIVE_ITEMS_ENABLE) ? save->strength_upgrade : MW_PROGRESSIVE_ITEMS_STATE[override.value.base.player].strength) {
        case 0: return GI_GORONS_BRACELET; // Goron Bracelet
        case 1: return GI_SILVER_GAUNTLETS; // Silver Gauntlets
        default: return GI_GOLD_GAUNTLETS; // Gold Gauntlets
    }
}

    // example 3
item_row_t *get_item_row(GetItemID item_id) {
    if (item_id >= GI_RANDO_MAX) {
        return NULL;
    }
    item_row_t *item_row = &(item_table[item_id]);
    if (item_row->base_item_id == 0) {
        return NULL;
    }
    return item_row;
}
```

</details>

### CODE REVIEW

File: ASM/c/item_table.h [hunk 1]
```diff
@@ -6,6 +6,326 @@
...
+
+    /* 0x00B8 */ GI_DOUBLE_DEFENSE, // Double Defense
+    /* 0x00B9 */ GI_MAGIC_METER, // Magic Meter
+    /* 0x00BA */ GI_DOUBLE_MAGIC, // Double Magic
+
+    /* 0x00BB */ GI_MINUET_OF_FOREST, // Minuet of Forest
+    /* 0x00BC */ GI_BOLERO_OF_FIRE, // Bolero of Fire
+    /* 0x00BD */ GI_SERENADE_OF_WATER, // Serenade of Water
+    /* 0x00BE */ GI_REQUIEM_OF_SPIRIT, // Requiem of Spirit
+    /* 0x00BF */ GI_NOCTURN_OF_SHADOW, // Nocturn of Shadow
```
Comment (justanr) [inline ASM/c/item_table.h]:
```suggestion
    /* 0x00BF */ GI_NOCTURNE_OF_SHADOW, // Nocturne of Shadow
```

File: ItemList.py [hunk 2]
```diff
@@ -1,6 +1,314 @@
...
+
+    GI_DOUBLE_DEFENSE = 0x00B8
+    GI_MAGIC_METER = 0x00B9
+    GI_DOUBLE_MAGIC = 0x00BA
+
+    GI_MINUET_OF_FOREST = 0x00BB
+    GI_BOLERO_OF_FIRE = 0x00BC
+    GI_SERENADE_OF_WATER = 0x00BD
+    GI_REQUIEM_OF_SPIRIT = 0x00BE
+    GI_NOCTURN_OF_SHADOW = 0x00BF
```
Comment (justanr) [inline ItemList.py]:
```suggestion
    GI_NOCTURNE_OF_SHADOW = 0x00BF
```

File: ASM/c/item_table.h [hunk 3]
```diff
@@ -6,6 +6,326 @@
...
+    /* 0x0077 */ GI_DEKU_STICK_UPGRADE_20, // Deku Stick Upgrade (20)
+    /* 0x0078 */ GI_DEKU_STICK_UPGRADE_30, // Deku Stick Upgrade (30)
+    /* 0x0079 */ GI_DEKU_NUT_UPGRADE_30, // Deku Nut Upgrade (30)
+    /* 0x007A */ GI_DEKU_NUT_UPGRADE_40, // Deku Nut Upgrade (40)
+    /* 0x007B */ GI_BULLET_BAG_50, // Bullet Bag (50)
+    /* 0x007C */ GI_ICE_TRAP, // Ice Trap
+    /* 0x007D */ GI_TEXT_0,
+    /* 0x007D */ GI_CAPPED_PIECE_OF_HEART = GI_TEXT_0, // Capped Piece of Heart
+    /* 0x007E */ GI_VANILLA_MAX,
+    /* 0x007E */ GI_CAPPED_HEART_CONTAINER = GI_VANILLA_MAX, // Capped Heart Container
```
Comment (cjohnson57) [inline ASM/c/item_table.h:144]: Can you explain this a little bit? I'm confused why the first 2 instances of the same ID are here

Comment (flagrama) [reply ASM/c/item_table.h:144]: Those are the vanilla values. GI_TEXT_0 is the pocket egg error id text that rando repurposes it seems and GI_VANILLLA_MAX is the size of the enum in vanilla.

Comment (cjohnson57) [reply ASM/c/item_table.h:144]: Ah ok, in that case it makes sense to document the original IDs as well as what we're using them for here.

**Comment by cjohnson57** [comment]
Looks good to me from a code perspective, I'd like to see someone play through a seed on this branch though, just to make sure everything's still working correctly

**Comment by fenhl** [comment]
I've played through most of a seed and everything seemed fine. So I'd say this should be good to merge after the upcoming release, a rebase or merge of latest Dev probably couldn't hurt though.

**Pull request merged**